# Capstone Project - Most Similar Neighboorhood

## Table of contents
* [Introduction: Business Problem](#introduction)
* [Data](#data)
* [Methodology](#methodology)
* [Analysis](#analysis)
* [Results and Discussion](#results)
* [Conclusion](#conclusion)

## Introduction: Business Problem <a name="introduction"></a>

In this project we will try to find the neighborhood similar to any city location in Bangalore City. Specifically, this report will be targeted to the people who are willing to relocate to bangalore and want to know which location in bangalore is having similar neighborhood in compare to the neighborhood they are staying.

Since there are lots of IT companies in the bangalore so employee from all over India is relocating here every months.Bangalore is the southern part of India and it's neighborhood(specially food resturants) are quite different from northern part. So, people who are relocating will prefer the neighborhood similar to the neighborhood they are staying now.

We will use our data science powers to generate a few most promissing neighborhoods based on this criteria. Advantages of each area will then be clearly expressed so that best possible final location can be chosen by the person.

## Data <a name="data"></a>

We decided to use locations(All SO & BO) centered around PinCodes of Bangalore, to define our neighborhoods.

Following data sources will be needed to extract/generate the required information:
* centers of candidate areas will be generated algorithmically and approximate addresses of centers of those areas will be obtained using **geopy**
* number of venues in the neighborhood and their type and location in every neighborhood will be obtained using **Foursquare API**
* coordinate of Bangalore center will be obtained using **geopy
*Details of person neighborhood will be  collected by user and coordinate is obtained using **geopy

In [1]:
import pandas as pd
!pip install lxml
!pip install geopy
from geopy.geocoders import Nominatim
import matplotlib.cm as cm
import matplotlib.colors as colors
import folium 
import requests
from pandas.io.json import json_normalize 

     |████████████████████████████████| 5.8MB 23.6MB/s eta 0:00:01     |█████████████████▌              | 3.1MB 23.6MB/s eta 0:00:01
     |████████████████████████████████| 112kB 7.1MB/s eta 0:00:01


### Loading data from https://finkode.com/ka/bangalore.html

In [2]:
html = 'https://finkode.com/ka/bangalore.html'
try: 
    bangalore_data = pd.read_html(html)[0]  
except IndexError:
    print("Table not found")
bangalore_data.rename(columns={"Post Office":"PostOffice",},inplace=True)
bangalore_data

PostOffice   District  Pincode
0       A F Station Yelahanka S.O  Bangalore   560063
1                     Adugodi S.O  Bangalore   560030
2                       Agara B.O  Bangalore   560034
3                       Agram S.O  Bangalore   560007
4                Amruthahalli B.O  Bangalore   560092
..                            ...        ...      ...
265                 Yelahanka S.O  Bangalore   560064
266  Yelahanka Satellite Town S.O  Bangalore   560064
267                   Yemalur B.O  Bangalore   560037
268        Yeshwanthpur Bazar S.O  Bangalore   560022
269              Yeswanthpura S.O  Bangalore   560022

[270 rows x 3 columns]

In [3]:
len(bangalore_data.Pincode.unique())


104

In [4]:
bangalore_data=bangalore_data.groupby('Pincode')['PostOffice'].apply(','.join).reset_index()


Pincode                                         PostOffice
0     560001  Bangalore Bazaar S.O,Bangalore G.P.O.,Cubban R...
1     560002  Bangalore City S.O,Bangalore Corporation Build...
2     560003  Malleswaram S.O,Palace Guttahalli S.O,Swimming...
3     560004  Basavanagudi H.O,Mavalli S.O,Pampamahakavi Roa...
4     560005                                    Fraser Town S.O
..       ...                                                ...
99    562120                                Chamarajasagara B.O
100   562125  Dommasandra B.O,Handenahalli B.O,Kugur B.O,Sar...
101   562130  Chikkanahalli B.O,Chunchanakuppe B.O,Kadabager...
102   562149  Bagalur S.O (Bangalore),Bandikodigehalli B.O,D...
103   562157  Bettahalsur S.O,Chikkajala B.O,Doddajala B.O,H...

[104 rows x 2 columns]

In [5]:
print(len(bangalore_data.Pincode.unique()))
bangalore_data['Latitude']=0
bangalore_data['Longitude']=0
print(bangalore_data.head())


104
   Pincode                                         PostOffice  Latitude  \
0   560001  Bangalore Bazaar S.O,Bangalore G.P.O.,Cubban R...         0   
1   560002  Bangalore City S.O,Bangalore Corporation Build...         0   
2   560003  Malleswaram S.O,Palace Guttahalli S.O,Swimming...         0   
3   560004  Basavanagudi H.O,Mavalli S.O,Pampamahakavi Roa...         0   
4   560005                                    Fraser Town S.O         0   

   Longitude  
0          0  
1          0  
2          0  
3          0  
4          0  


### Adding latitude & longitude of pincodes

In [6]:
def lat_lng(bangalore_data):
    for i in range(len(bangalore_data)):
        geolocator = Nominatim(user_agent="ny_explorer")
        location = geolocator.geocode(bangalore_data.iloc[i,0] )
        if location != None:
            bangalore_data.iloc[i,2] = location.latitude
            bangalore_data.iloc[i,3] = location.longitude
    return bangalore_data
    

In [7]:
fin_bangalore_data=lat_lng(bangalore_data)

In [8]:
fin_bangalore_data

Pincode                                         PostOffice   Latitude  \
0     560001  Bangalore Bazaar S.O,Bangalore G.P.O.,Cubban R... -33.038136   
1     560002  Bangalore City S.O,Bangalore Corporation Build...  12.958625   
2     560003  Malleswaram S.O,Palace Guttahalli S.O,Swimming...  13.000240   
3     560004  Basavanagudi H.O,Mavalli S.O,Pampamahakavi Roa...  12.944829   
4     560005                                    Fraser Town S.O  13.428852   
..       ...                                                ...        ...   
99    562120                                Chamarajasagara B.O  12.922515   
100   562125  Dommasandra B.O,Handenahalli B.O,Kugur B.O,Sar...  12.887615   
101   562130  Chikkanahalli B.O,Chunchanakuppe B.O,Kadabager...  13.855864   
102   562149  Bagalur S.O (Bangalore),Bandikodigehalli B.O,D...  13.116054   
103   562157  Bettahalsur S.O,Chikkajala B.O,Doddajala B.O,H...  13.226201   

      Longitude  
0    137.575919  
1     77.577567  
2     77.565249  
3     77.567095  
4     77.627203  
..          ...  
99    77.215788  
100   77.739003  
101   75.960947  
102   77.667895  
103   77.527476  

[104 rows x 4 columns]

In [182]:
clustering_data=fin_bangalore_data[fin_bangalore_data.Latitude>0]

In [183]:
clustering_data.shape

(101, 4)

In [185]:
clustering_data.head()

Pincode                                         PostOffice   Latitude  \
1   560002  Bangalore City S.O,Bangalore Corporation Build...  12.958625   
2   560003  Malleswaram S.O,Palace Guttahalli S.O,Swimming...  13.000240   
3   560004  Basavanagudi H.O,Mavalli S.O,Pampamahakavi Roa...  12.944829   
4   560005                                    Fraser Town S.O  13.428852   
5   560006             J.C.Nagar S.O,Training Command IAF S.O  13.006087   

   Longitude  
1  77.577567  
2  77.565249  
3  77.567095  
4  77.627203  
5  77.593151

In [186]:
address = 'Bangalore'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Bangalore City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Bangalore City are 12.9791198, 77.5912997.


### Plotting the map of pincodes of bangalore

In [187]:
 
# create map of Manhattan using latitude and longitude values
map_to = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(clustering_data['Latitude'], clustering_data['Longitude'], clustering_data['PostOffice']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_to)  
    
map_to

### FourSuareAPI for finding nearby venues

In [188]:
CLIENT_ID = 'WH4VWFY0GYS4CYZPF3CLQSHRNSJKERAWCHP5YRKPXUQ4MAYZ' # your Foursquare ID
CLIENT_SECRET = 'UOBCXK5OCVXJYS1VXHANPT5OL2QRJRVBUEHN0I5EJPVSMU4W' # your Foursquare Secret
VERSION = '20180605'  # Foursquare API version
neighborhoods_bangalore=clustering_data
neighborhoods_bangalore.rename(columns={"PostOffice":"Neighborhood"},inplace=True)


/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/pandas/core/frame.py:4133: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [189]:
def getNearbyVenues(names, latitudes, longitudes, radius=500,LIMIT = 100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        #print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [190]:
bangalore_venues = getNearbyVenues(names=neighborhoods_bangalore['Neighborhood'],
                                   latitudes=neighborhoods_bangalore['Latitude'],
                                   longitudes=neighborhoods_bangalore['Longitude']
                                  )

print(bangalore_venues.shape)
bangalore_venues

(862, 7)


Neighborhood  Neighborhood Latitude  \
0    Bangalore City S.O,Bangalore Corporation Build...              12.958625   
1    Bangalore City S.O,Bangalore Corporation Build...              12.958625   
2    Bangalore City S.O,Bangalore Corporation Build...              12.958625   
3    Bangalore City S.O,Bangalore Corporation Build...              12.958625   
4    Malleswaram S.O,Palace Guttahalli S.O,Swimming...              13.000240   
..                                                 ...                    ...   
857                Bangalore International Airport S.O              13.199828   
858                Bangalore International Airport S.O              13.199828   
859                Bangalore International Airport S.O              13.199828   
860  Attibele S.O,Bidaraguppe B.O,Mayasandra B.O,Ne...              12.777538   
861  Bagalur S.O (Bangalore),Bandikodigehalli B.O,D...              13.116054   

     Neighborhood Longitude                  Venue  Venue Latitude  \
0                 77.577567       NMH Tiffin House       12.954300   
1                 77.577567      Mayuri Restaurant       12.956120   
2                 77.577567         Hotel Nandhini       12.955307   
3                 77.577567         Bangalore Fort       12.962529   
4                 77.565249     Raghavendra Stores       13.000799   
..                      ...                    ...             ...   
857               77.703860      Bangalore Run Way       13.201992   
858               77.703860              Windmills       13.197953   
859               77.703860  Thai Airways check-in       13.198118   
860               77.765040      Shree Ranga Vilas       12.779921   
861               77.667895           goal station       13.117133   

     Venue Longitude           Venue Category  
0          77.578806        Indian Restaurant  
1          77.580845        Indian Restaurant  
2          77.579498        Indian Restaurant  
3          77.575816            Historic Site  
4          77.563924           Breakfast Spot  
..               ...                      ...  
857        77.701385            Moving Target  
858        77.707560                 Beer Bar  
859        77.707842          Airport Service  
860        77.768066  South Indian Restaurant  
861        77.670192              Sports Club  

[862 rows x 7 columns]

Now we have all the collected the pincodes data it's latitude,longitude , top nearby venues  and it's coordinates we are good to further.
This concludes the data gathering phase - we're now ready to use this data for analysis to produce the report on most likely neighborhood.

## Methodology <a name="methodology"></a>

In this project we will direct our efforts on detecting areas of Bangalore that have neighbrhood similar to the users neighborhood.

In first step we have collected the required **data: pincode, areas in the pincode as neighborhood,top 10 venues nearby the pincode(according to Foursquare categorization).

Second step in our analysis will be calculation and exploration of 'neighborhood' across different areas of bangalore - ,what are the most frequent venues per pincode,doing proper encoding of data, and then applying k means clustering on the top of it with k=5. That means we are going to divide our entire bangalore area in 5 clusters.

In third and final step we will focus on the most similar area as per the user input and we will take the current place and current pincode as input from user and then we will identify the areas in the bangalore which is most similar to the one specifed by user. 

## Analysis <a name="analysis"></a>

Let's perform some basic explanatory data analysis and derive some additional info from our raw data. First let's count the **number of venues in every neighborhood**:

In [191]:
bangalore_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
A F Station Yelahanka S.O,BSF Campus Yelahanka S.O                      1   
Adugodi S.O                                                             4   
Agram S.O                                                               4   
Amruthahalli B.O,Byatarayanapura B.O,Kodigehall...                     12   
Anandnagar S.O (Bangalore),H.A. Farm S.O,Hebbal...                      7   
...                                                                   ...   
Science Institute S.O                                                   1   
Seshadripuram S.O                                                      39   
Sivan Chetty Gardens S.O                                               13   
Tarabanahalli B.O                                                       2   
Yeshwanthpur Bazar S.O,Yeswanthpura S.O                                 6   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
A F Station Yelahanka S.O,BSF Campus Yelahanka S.O                       1   
Adugodi S.O                                                              4   
Agram S.O                                                                4   
Amruthahalli B.O,Byatarayanapura B.O,Kodigehall...                      12   
Anandnagar S.O (Bangalore),H.A. Farm S.O,Hebbal...                       7   
...                                                                    ...   
Science Institute S.O                                                    1   
Seshadripuram S.O                                                       39   
Sivan Chetty Gardens S.O                                                13   
Tarabanahalli B.O                                                        2   
Yeshwanthpur Bazar S.O,Yeswanthpura S.O                                  6   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
A F Station Yelahanka S.O,BSF Campus Yelahanka S.O      1               1   
Adugodi S.O                                             4               4   
Agram S.O                                               4               4   
Amruthahalli B.O,Byatarayanapura B.O,Kodigehall...     12              12   
Anandnagar S.O (Bangalore),H.A. Farm S.O,Hebbal...      7               7   
...                                                   ...             ...   
Science Institute S.O                                   1               1   
Seshadripuram S.O                                      39              39   
Sivan Chetty Gardens S.O                               13              13   
Tarabanahalli B.O                                       2               2   
Yeshwanthpur Bazar S.O,Yeswanthpura S.O                 6               6   

                                                    Venue Longitude  \
Neighborhood                                                          
A F Station Yelahanka S.O,BSF Campus Yelahanka S.O                1   
Adugodi S.O                                                       4   
Agram S.O                                                         4   
Amruthahalli B.O,Byatarayanapura B.O,Kodigehall...               12   
Anandnagar S.O (Bangalore),H.A. Farm S.O,Hebbal...                7   
...                                                             ...   
Science Institute S.O                                             1   
Seshadripuram S.O                                                39   
Sivan Chetty Gardens S.O                                         13   
Tarabanahalli B.O                                                 2   
Yeshwanthpur Bazar S.O,Yeswanthpura S.O                           6   

                                                    Venue Category  
Neighborhood                            

Let's look at the unique venue categories

In [192]:
print('There are {} uniques categories.'.format(len(bangalore_venues['Venue Category'].unique())))

There are 165 uniques categories.


### Perform one hot encoding on the data

In [193]:
#one hot encoding
bangalore_onehot = pd.get_dummies(bangalore_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
bangalore_onehot['Neighborhood'] = bangalore_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [bangalore_onehot.columns[-1]] + list(bangalore_onehot.columns[:-1])
bangalore_onehot = bangalore_onehot[fixed_columns]

bangalore_onehot.head()

Neighborhood  ATM  Accessories Store  \
0  Bangalore City S.O,Bangalore Corporation Build...    0                  0   
1  Bangalore City S.O,Bangalore Corporation Build...    0                  0   
2  Bangalore City S.O,Bangalore Corporation Build...    0                  0   
3  Bangalore City S.O,Bangalore Corporation Build...    0                  0   
4  Malleswaram S.O,Palace Guttahalli S.O,Swimming...    0                  0   

   Afghan Restaurant  Airport Service  American Restaurant  Andhra Restaurant  \
0                  0                0                    0                  0   
1                  0                0                    0                  0   
2                  0                0                    0                  0   
3                  0                0                    0                  0   
4                  0                0                    0                  0   

   Arcade  Art Gallery  Arts & Crafts Store  ...  Toy / Game Store  Track  \
0       0            0                    0  ...                 0      0   
1       0            0                    0  ...                 0      0   
2       0            0                    0  ...                 0      0   
3       0            0                    0  ...                 0      0   
4       0            0                    0  ...                 0      0   

   Trail  Train Station  Udupi Restaurant  Vegetarian / Vegan Restaurant  \
0      0              0                 0                              0   
1      0              0                 0                              0   
2      0              0                 0                              0   
3      0              0                 0                              0   
4      0              0                 0                              0   

   Vietnamese Restaurant  Waterfront  Wine Shop  Women's Store  
0                      0           0          0              0  
1                      0           0          0              0  
2                      0           0          0              0  
3                      0           0          0              0  
4                      0           0          0              0  

[5 rows x 166 columns]

In [194]:
bangalore_grouped = bangalore_onehot.groupby('Neighborhood').mean().reset_index()
bangalore_grouped

Neighborhood  ATM  Accessories Store  \
0   A F Station Yelahanka S.O,BSF Campus Yelahanka...  0.0                0.0   
1                                         Adugodi S.O  0.0                0.0   
2                                           Agram S.O  0.0                0.0   
3   Amruthahalli B.O,Byatarayanapura B.O,Kodigehal...  0.0                0.0   
4   Anandnagar S.O (Bangalore),H.A. Farm S.O,Hebba...  0.0                0.0   
..                                                ...  ...                ...   
83                              Science Institute S.O  0.0                0.0   
84                                  Seshadripuram S.O  0.0                0.0   
85                           Sivan Chetty Gardens S.O  0.0                0.0   
86                                  Tarabanahalli B.O  0.0                0.0   
87            Yeshwanthpur Bazar S.O,Yeswanthpura S.O  0.0                0.0   

    Afghan Restaurant  Airport Service  American Restaurant  \
0                 0.0              0.0             0.000000   
1                 0.0              0.0             0.000000   
2                 0.0              0.0             0.000000   
3                 0.0              0.0             0.000000   
4                 0.0              0.0             0.000000   
..                ...              ...                  ...   
83                0.0              0.0             0.000000   
84                0.0              0.0             0.025641   
85                0.0              0.0             0.000000   
86                0.0              0.0             0.000000   
87                0.0              0.0             0.000000   

    Andhra Restaurant    Arcade  Art Gallery  Arts & Crafts Store  ...  \
0                 0.0  0.000000          0.0                  0.0  ...   
1                 0.0  0.000000          0.0                  0.0  ...   
2                 0.0  0.000000          0.0                  0.0  ...   
3                 0.0  0.000000          0.0                  0.0  ...   
4                 0.0  0.000000          0.0                  0.0  ...   
..                ...       ...          ...                  ...  ...   
83                0.0  0.000000          0.0                  0.0  ...   
84                0.0  0.051282          0.0                  0.0  ...   
85                0.0  0.000000          0.0                  0.0  ...   
86                0.0  0.000000          0.0                  0.0  ...   
87                0.0  0.000000          0.0                  0.0  ...   

    Toy / Game Store  Track  Trail  Train Station  Udupi Restaurant  \
0                0.0    0.0    0.0            0.0               0.0   
1                0.0    0.0    0.0            0.0               0.0   
2                0.0    0.0    0.0            0.0               0.0   
3                0.0    0.0    0.0            0.0               0.0   
4                0.0    0.0    0.0            0.0               0.0   
..               ...    ...    ...            ...               ...   
83               0.0    0.0    0.0            0.0               0.0   
84               0.0    0.0    0.0            0.0               0.0   
85               0.0    0.0    0.0            0.0               0.0   
86               0.0    0.0    0.0            0.0               0.0   
87               0.0    0.0    0.0            0.0               0.0   

    Vegetarian / Vegan Restaurant  Vietnamese Restaurant  Waterfront  \
0                             0.0                    0.0         0.0   
1                             0.0                    0.0         0.0   
2                             0.0                    0.0         0.0   
3                             0.0                    0.0         0.0   
4                             0.0                    0.0         0.0   
..                            ...                    ...         ...   
83                            0.0                    0.0         

In [195]:
bangalore_grouped.shape

(88, 166)

Let's checkout the top 5 venues in each neighborhood along with it's avergae frequency

In [196]:
num_top_venues = 5

for hood in bangalore_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = bangalore_grouped[bangalore_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----A F Station Yelahanka S.O,BSF Campus Yelahanka S.O----
             venue  freq
0        Juice Bar   1.0
1              ATM   0.0
2             Park   0.0
3  Motorcycle Shop   0.0
4    Movie Theater   0.0


----Adugodi S.O----
         venue  freq
0   Playground  0.25
1       Bakery  0.25
2  Pizza Place  0.25
3         Café  0.25
4          ATM  0.00


----Agram S.O----
            venue  freq
0          Casino  0.25
1       Juice Bar  0.25
2     Pizza Place  0.25
3  Breakfast Spot  0.25
4             ATM  0.00


----Amruthahalli B.O,Byatarayanapura B.O,Kodigehalli B.O,Sahakaranagar P.O S.O----
               venue  freq
0  Indian Restaurant  0.17
1  Indian Sweet Shop  0.08
2       Liquor Store  0.08
3             Resort  0.08
4     Sandwich Place  0.08


----Anandnagar S.O (Bangalore),H.A. Farm S.O,Hebbal Kempapura S.O----
               venue  freq
0  Indian Restaurant  0.29
1             Market  0.14
2        Coffee Shop  0.14
3        Pizza Place  0.14
4   Department Store  0.1

Let's try to transform the data to in more better format in which we can easily visulaize top 10 categories of each neighborhood.

In [197]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [198]:
import numpy as np
num_top_venues = 50

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = bangalore_grouped['Neighborhood']

for ind in np.arange(bangalore_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(bangalore_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0  A F Station Yelahanka S.O,BSF Campus Yelahanka...             Juice Bar   
1                                        Adugodi S.O            Playground   
2                                          Agram S.O                Casino   
3  Amruthahalli B.O,Byatarayanapura B.O,Kodigehal...     Indian Restaurant   
4  Anandnagar S.O (Bangalore),H.A. Farm S.O,Hebba...     Indian Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0         Women's Store                 Diner     Fish & Chips Shop   
1           Pizza Place                Bakery                  Café   
2           Pizza Place             Juice Bar        Breakfast Spot   
3                Resort       Badminton Court        Sandwich Place   
4      Department Store           Coffee Shop           Pizza Place   

        5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0  Financial or Legal Service  Fast Food Restaurant        Farmers Market   
1                       Diner  Fast Food Restaurant        Farmers Market   
2               Women's Store                 Diner  Fast Food Restaurant   
3                        Café    Italian Restaurant          Liquor Store   
4                        Park                Market                 Diner   

  8th Most Common Venue 9th Most Common Venue  ... 41th Most Common Venue  \
0           Event Space     Electronics Store  ...         Clothing Store   
1           Event Space     Electronics Store  ...      Food & Drink Shop   
2        Farmers Market           Event Space  ...      Food & Drink Shop   
3     Indian Sweet Shop           Snack Place  ...             Food Court   
4  Fast Food Restaurant        Farmers Market  ...      Food & Drink Shop   

  42th Most Common Venue 43th Most Common Venue 44th Most Common Venue  \
0     Chinese Restaurant         Ice Cream Shop        Bed & Breakfast   
1            Flea Market     Chinese Restaurant         Chocolate Shop   
2           Cocktail Bar     Chinese Restaurant         Chocolate Shop   
3      Food & Drink Shop         Clothing Store            Flea Market   
4         Clothing Store         Chocolate Shop     Chinese Restaurant   

  45th Most Common Venue 46th Most Common Venue 47th Most Common Venue  \
0                 Bakery        Badminton Court              BBQ Joint   
1              Hotel Bar                    Bar        Badminton Court   
2              Hotel Bar                 Bakery        Badminton Court   
3           Cocktail Bar            Chaat Place         Chocolate Shop   
4        Bed & Breakfast                 Bakery        Badminton Court   

  48th Most Common Venue 49th Most Common Venue 50th Most Common Venue  
0          Auto Workshop            Auto Garage     Athletics & Sports  
1              BBQ Joint          Auto Workshop            Auto Garage  
2              BBQ Joint          Auto Workshop            Auto Garage  
3     Chinese Restaurant                    Bar                 Bakery  
4              BBQ Joint          Auto Workshop            Auto Garage  

[5 rows x 51 columns]

## Clustering the neighborhood (K Means Clustering)

In [199]:
from sklearn.cluster import KMeans
# set number of clusters
kclusters = 5

bangalore_grouped_clustering = bangalore_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(bangalore_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] # add clustering labels


array([1, 1, 1, 2, 2, 2, 1, 1, 1, 0], dtype=int32)

In [200]:
# add clustering labels

neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
bangalore_merged = neighborhoods_bangalore


bangalore_merged = bangalore_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')
bangalore_merged.head() # check the last columns!

Pincode                                       Neighborhood   Latitude  \
1   560002  Bangalore City S.O,Bangalore Corporation Build...  12.958625   
2   560003  Malleswaram S.O,Palace Guttahalli S.O,Swimming...  13.000240   
3   560004  Basavanagudi H.O,Mavalli S.O,Pampamahakavi Roa...  12.944829   
4   560005                                    Fraser Town S.O  13.428852   
5   560006             J.C.Nagar S.O,Training Command IAF S.O  13.006087   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
1  77.577567             2.0     Indian Restaurant         Historic Site   
2  77.565249             1.0              Pharmacy      Department Store   
3  77.567095             2.0     Indian Restaurant  Fast Food Restaurant   
4  77.627203             NaN                   NaN                   NaN   
5  77.593151             1.0         Boat or Ferry         Bowling Alley   

  3rd Most Common Venue 4th Most Common Venue       5th Most Common Venue  \
1         Women's Store                 Diner  Financial or Legal Service   
2           Snack Place                Bakery              Breakfast Spot   
3           Pizza Place           Art Gallery              Farmers Market   
4                   NaN                   NaN                         NaN   
5                  Café           Auto Garage               Women's Store   

   ... 41th Most Common Venue 42th Most Common Venue 43th Most Common Venue  \
1  ...         Clothing Store     Chinese Restaurant         Ice Cream Shop   
2  ...       Football Stadium             Food Court      Food & Drink Shop   
3  ...             Food Truck             Food Court           Cocktail Bar   
4  ...                    NaN                    NaN                    NaN   
5  ...             Food Truck             Food Court           Cocktail Bar   

  44th Most Common Venue 45th Most Common Venue 46th Most Common Venue  \
1        Bed & Breakfast                 Bakery        Badminton Court   
2                   Food         Chocolate Shop            Flea Market   
3      Food & Drink Shop     Chinese Restaurant         Clothing Store   
4                    NaN                    NaN                    NaN   
5     Chinese Restaurant         Chocolate Shop   Arts & Entertainment   

  47th Most Common Venue 48th Most Common Venue 49th Most Common Venue  \
1              BBQ Joint          Auto Workshop            Auto Garage   
2                 Casino     Chinese Restaurant   Arts & Entertainment   
3         Chocolate Shop               Beer Bar        Bed & Breakfast   
4                    NaN                    NaN                    NaN   
5                 Bakery        Badminton Court              BBQ Joint   

  50th Most Common Venue  
1     Athletics & Sports  
2        Badminton Court  
3                    Bar  
4                    NaN  
5          Auto Workshop  

[5 rows x 55 columns]

In [201]:
bangalore_merged = bangalore_merged.dropna()
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(bangalore_merged['Latitude'], bangalore_merged['Longitude'], bangalore_merged['Neighborhood'], bangalore_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Let's have a look at the 5 clusters formed by Kmeans

In [202]:
bangalore_merged.loc[bangalore_merged['Cluster Labels'] == 0, bangalore_merged.columns[[1] + list(range(5, bangalore_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
29  Kanakanagar S.O,P&T Col. Kavalbyrasandra S.O,R...  Fast Food Restaurant   
43         Austin Town S.O,Viveknagar S.O (Bangalore)   Indie Movie Theater   
55                                    Chudenapura B.O           Bus Station   
71       Dr. Shivarama Karanth Nagar S.O,Kothanur S.O                   ATM   
75                                         Bolare B.O        Boarding House   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
29                Resort           Supermarket                  Park   
43         Historic Site      Football Stadium           Bus Station   
55            Restaurant         Women's Store                 Diner   
71             Bookstore           Bus Station             Wine Shop   
75   Indie Movie Theater            Restaurant           Bus Station   

         5th Most Common Venue        6th Most Common Venue  \
29                 Bus Station                Women's Store   
43               Women's Store                        Diner   
55  Financial or Legal Service         Fast Food Restaurant   
71             Airport Service  Eastern European Restaurant   
75                Tennis Court                Women's Store   

   7th Most Common Venue 8th Most Common Venue       9th Most Common Venue  \
29                 Diner        Farmers Market                 Event Space   
43  Fast Food Restaurant        Farmers Market                 Event Space   
55        Farmers Market           Event Space           Electronics Store   
71           Flea Market     Fish & Chips Shop  Financial or Legal Service   
75                 Diner  Fast Food Restaurant              Farmers Market   

    ... 41th Most Common Venue 42th Most Common Venue 43th Most Common Venue  \
29  ...             Food Court      Food & Drink Shop         Clothing Store   
43  ...      Food & Drink Shop         Clothing Store     Chinese Restaurant   
55  ...             Food Court         Clothing Store            Chaat Place   
71  ...    Fried Chicken Joint       Football Stadium            Coffee Shop   
75  ...             Food Truck             Food Court      Food & Drink Shop   

   44th Most Common Venue 45th Most Common Venue 46th Most Common Venue  \
29            Chaat Place     Chinese Restaurant              Hotel Bar   
43         Ice Cream Shop            Chaat Place                 Bakery   
55     Chinese Restaurant         Ice Cream Shop                 Bakery   
71         Clothing Store    American Restaurant            Auto Garage   
75         Clothing Store            Chaat Place     Chinese Restaurant   

   47th Most Common Venue 48th Most Common Venue 49th Most Common Venue  \
29                 Bakery        Badminton Court              BBQ Joint   
43        Badminton Court              BBQ Joint          Auto Workshop   
55        Badminton Court              BBQ Joint          Auto Workshop   
71               Beer Bar        Bed & Breakfast                    Bar   
75   Arts & Entertainment        Badminton Court              BBQ Joint   

   50th Most Common Venue  
29          Auto Workshop  
43            Auto Garage  
55            Auto Garage  
71                 Bakery  
75          Auto Workshop  

[5 rows x 51 columns]

In [203]:
bangalore_merged.loc[bangalore_merged['Cluster Labels'] == 1, bangalore_merged.columns[[1] + list(range(5, bangalore_merged.shape[1]))]]

Neighborhood  \
2    Malleswaram S.O,Palace Guttahalli S.O,Swimming...   
5               J.C.Nagar S.O,Training Command IAF S.O   
6                                            Agram S.O   
7                 H.A.L II Stage H.O,Hulsur Bazaar S.O   
11                               Science Institute S.O   
14                                  Jalahalli West S.O   
15   Doorvaninagar S.O,Krishnarajapuram R S S.O,Ram...   
18   Gaviopuram Extension S.O,Narasimharaja Colony S.O   
19                                   Seshadripuram S.O   
22                                     Magadi Road S.O   
24   CMP Centre And School S.O,Museum Road S.O,Rich...   
28                                         Adugodi S.O   
33                 Devasandra S.O,Krishnarajapuram S.O   
34   Doddanekkundi B.O,Kundalahalli S.O,Marathahall...   
36                                    Nayandahalli S.O   
42                                     Benson Town S.O   
44                          Hoodi B.O,Mahadevapura S.O   
46   Ashoknagar S.O (Bangalore),Banashankari S.O,Da...   
49        Mathikere S.O,Msrit S.O,Sri Chowdeshwari S.O   
50                Malleswaram West S.O,Milk Colony S.O   
54                                      Rv Niketan S.O   
57   Doddakallasandra S.O,Konanakunte B.O,Tataguni B.O   
58   A F Station Yelahanka S.O,BSF Campus Yelahanka...   
59   Attur B.O,CRPF Campus Yelahanka S.O,Jakkur B.O...   
60                                        G.K.V.K. S.O   
62   Devanagundi B.O,Kadugodi S.O,Kannamangala B.O,...   
64   B Sk II Stage S.O,Jayanagar West S.O,Padmanabh...   
66                                      Nagarbhavi S.O   
70   Bannerghatta Road S.O,Bilekahalli B.O,Hulimavu...   
72   J P Nagar S.O,JP Nagar III Phase S.O,Kumaraswa...   
78   Banashankari III Stage S.O,Byatarayanapura B.O...   
81                                   Tarabanahalli B.O   
87               Kanteeravanagar S.O,Nandinilayout S.O   
89           Kenchanahalli B.O,Rajarajeshwarinagar S.O   
91                                Electronics City S.O   
93                          Bellandur S.O,Panathur B.O   
95   Haragadde B.O,Hennagara B.O,Hulimangala B.O,Ji...   
96                 Bangalore International Airport S.O   
98   Attibele S.O,Bidaraguppe B.O,Mayasandra B.O,Ne...   
102  Bagalur S.O (Bangalore),Bandikodigehalli B.O,D...   

       1st Most Common Venue          2nd Most Common Venue  \
2                   Pharmacy               Department Store   
5              Boat or Ferry                  Bowling Alley   
6                     Casino                    Pizza Place   
7          Indian Restaurant                            Pub   
11                 Bookstore                  Women's Store   
14      Fast Food Restaurant                      BBQ Joint   
15                   Stadium                       Platform   
18               Pizza Place           Fast Food Restaurant   
19         Indian Restaurant                 Clothing Store   
22            Tennis Stadium           Karnataka Restaurant   
24                      Café              Indian Restaurant   
28                Playground                    Pizza Place   
33                       ATM              Convenience Store   
34                     Hotel                     Restaurant   
36               Flea Market              Accessories Store   
42               Music Venue              Indian Restaurant   
44               Coffee Shop                           Café   
46      Fast Food Restaurant               Asian Restaurant   
49                     Diner                         Bakery   
50            Ice Cream Shop               Department Store   
54            Sandwich Place                  Women's Store   
57         Afghan Restaurant        Fruit & Vegetable Store   
58                 Juice Bar                  Women's Store   
59              Dessert Shop                    Pizza Place   
60              Dessert Shop        Fruit & Vegetable Store   
62                  Platf

In [204]:
bangalore_merged.loc[bangalore_merged['Cluster Labels'] == 2, bangalore_merged.columns[[1] + list(range(5, bangalore_merged.shape[1]))]]

Neighborhood  \
1   Bangalore City S.O,Bangalore Corporation Build...   
3   Basavanagudi H.O,Mavalli S.O,Pampamahakavi Roa...   
8      Bangalore Dist Offices Bldg S.O,K. G. Road S.O   
9   Industrial Estate S.O (Bangalore),Rajajinagar ...   
12                                      Jalahalli H.O   
16                             NAL S.O,Vimanapura S.O   
17                         Chamrajpet S.O (Bangalore)   
20                  Gayathrinagar S.O,Srirampuram S.O   
21            Yeshwanthpur Bazar S.O,Yeswanthpura S.O   
23  Anandnagar S.O (Bangalore),H.A. Farm S.O,Hebba...   
26  Sampangiramnagar S.O,Shanthinagar S.O,Wilson G...   
27               Dharmaram College S.O,Taverekere B.O   
30            Malkand Lines S.O,Maruthi Sevanagar S.O   
32                                     Carmelaram S.O   
35                        Indiranagar S.O (Bangalore)   
37  Chandra Lay Out S.O,Vijayanagar East S.O,Vijay...   
38           Jayanagar H.O,Tilaknagar S.O (Bangalore)   
39                           Sivan Chetty Gardens S.O   
40  Banaswadi S.O,Horamavu B.O,Jalavayuvihar S.O,K...   
41  Arabic College S.O,Nagawara B.O,Venkateshapura...   
47                                    H.K.P. Road S.O   
48                                       Chickpet S.O   
51  Bnagalore Viswavidalaya S.O,Mallathahalli B.O,...   
52                             Peenya Dasarahalli S.O   
56              Chikkalasandra S.O,Subramanyapura S.O   
61  Immedihalli B.O,Ramagondana Halli B.O,Whitefie...   
63  Begur B.O,Bommanahalli S.O (Bangalore),Hongasa...   
65                                         Domlur S.O   
68  Kumbalagodu S.O,Kumbalgodu Gollahalli B.O,Ramo...   
73  Basaveshwaranagar S.O,K H B Colony S.O,Kamaksh...   
74                                 Sadashivanagar S.O   
76  Bannerghatta S.O,Gottigere S.O,Kallubalu B.O,R...   
77  Kacharakanahalli B.O,Lingarajapuram S.O,St. Th...   
79  Mahalakshmipuram Layout S.O,West of Chord Road...   
83  Amruthahalli B.O,Byatarayanapura B.O,Kodigehal...   
84                                C.V.Raman Nagar S.O   
85  ISRO Anthariksha Bhavan S.O,R.M.V. Extension I...   
86                              Koramangala VI Bk S.O   
92                                     HSR Layout S.O   
94                                     Hampinagar S.O   

            1st Most Common Venue          2nd Most Common Venue  \
1               Indian Restaurant                  Historic Site   
3               Indian Restaurant           Fast Food Restaurant   
8               Indian Restaurant                          Hotel   
9               Indian Restaurant                         Bakery   
12              Indian Restaurant           Fast Food Restaurant   
16              Indian Restaurant                     Restaurant   
17          General Entertainment              Indian Restaurant   
20              Indian Restaurant                         Bakery   
21           Fast Food Restaurant             Miscellaneous Shop   
23              Indian Restaurant               Department Store   
26  Vegetarian / Vegan Restaurant              Indian Restaurant   
27              Indian Restaurant               Asian Restaurant   
30              Indian Restaurant              Andhra Restaurant   
32              Indian Restaurant                  Women's Store   
35              Indian Restaurant                           Café   
37              Indian Restaurant                    Bus Station   
38              Indian Restaurant               Department Store   
39              Indian Restaurant                           Café   
40              Indian Restaurant      Indian Chinese Restaurant   
41              Indian Restaurant                 Clothing Store   
47              Indian Restaurant                           Café   
48              Indian Restaurant             Miscellaneous Shop   
51                            ATM              Indian Restaurant   
52              Indian Restaurant                    Bus Station   
56 

In [205]:
bangalore_merged.loc[bangalore_merged['Cluster Labels'] == 3, bangalore_merged.columns[[1] + list(range(5, bangalore_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
13                                 Jalahalli East S.O                   ATM   
82  Bapagrama B.O,Herohalli B.O,Herohalli S.O,Visw...                   ATM   

   2nd Most Common Venue       3rd Most Common Venue 4th Most Common Venue  \
13           Flea Market  Financial or Legal Service  Fast Food Restaurant   
82            Astrologer                  Donut Shop     Fish & Chips Shop   

         5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
13              Farmers Market           Event Space     Electronics Store   
82  Financial or Legal Service  Fast Food Restaurant        Farmers Market   

          8th Most Common Venue 9th Most Common Venue  ...  \
13  Eastern European Restaurant            Donut Shop  ...   
82                  Event Space     Electronics Store  ...   

   41th Most Common Venue 42th Most Common Venue 43th Most Common Venue  \
13         Chocolate Shop     Chinese Restaurant            Chaat Place   
82         Clothing Store     Chinese Restaurant         Ice Cream Shop   

   44th Most Common Venue 45th Most Common Venue 46th Most Common Venue  \
13                 Casino                 Bakery        Badminton Court   
82            Chaat Place                    Bar                 Bakery   

   47th Most Common Venue 48th Most Common Venue 49th Most Common Venue  \
13              BBQ Joint          Auto Workshop            Auto Garage   
82        Badminton Court              BBQ Joint          Auto Workshop   

   50th Most Common Venue  
13     Athletics & Sports  
82            Auto Garage  

[2 rows x 51 columns]

In [206]:
bangalore_merged.loc[bangalore_merged['Cluster Labels'] == 4, bangalore_merged.columns[[1] + list(range(5, bangalore_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
90  Bommasandra Industrial Estate S.O,Chandapura B...                Bakery   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
90         Women's Store            Donut Shop     Fish & Chips Shop   

         5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
90  Financial or Legal Service  Fast Food Restaurant        Farmers Market   

   8th Most Common Venue 9th Most Common Venue  ... 41th Most Common Venue  \
90           Event Space     Electronics Store  ...         Clothing Store   

   42th Most Common Venue 43th Most Common Venue 44th Most Common Venue  \
90     Chinese Restaurant         Ice Cream Shop            Chaat Place   

   45th Most Common Venue 46th Most Common Venue 47th Most Common Venue  \
90                    Bar        Badminton Court              BBQ Joint   

   48th Most Common Venue 49th Most Common Venue 50th Most Common Venue  
90          Auto Workshop            Auto Garage     Athletics & Sports  

[1 rows x 51 columns]

So,our bangalore neighborhood model is ready now we are going to take a demo user input and do proper transformation and then we will predict which cluster it belongs to and the pincodes/neighborhood in that cluster will be considered as suggestion for similar neighborhood in bangalore.

## Suggesting Neighborhood

In [286]:
# current_place='Chandni Chowk'
# current_pincode=110006
# current_city='Delhi'

current_place='Sarojini Nagar'
current_pincode=110023
current_city='Delhi'



### loading test data in proper format

In [287]:
def create_test_data(current_place,current_pincode,current_city):
    List = [[current_place, current_pincode]]
    predict_neighborhood=pd.DataFrame(List )
    predict_neighborhood.rename(columns={0:"Neighborhood",1:"Pincode"},inplace=True)
    predict_neighborhood['Latitude']=0
    predict_neighborhood['Longitude']=0
    predict_neighborhood=lat_lng(predict_neighborhood)
    print(predict_neighborhood.head())
    return predict_neighborhood


In [288]:
predict_neighborhood=create_test_data(current_place,current_pincode,current_city)

     Neighborhood  Pincode   Latitude  Longitude
0  Sarojini Nagar   110023  28.574157   77.19537


### Getting nearby venues of test data

In [289]:
predict_venues = getNearbyVenues(names=predict_neighborhood['Neighborhood'],
                                   latitudes=predict_neighborhood['Latitude'],
                                   longitudes=predict_neighborhood['Longitude']
                                  )

print(predict_venues.shape)
predict_venues.head()

(12, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Sarojini Nagar              28.574157                77.19537   
1  Sarojini Nagar              28.574157                77.19537   
2  Sarojini Nagar              28.574157                77.19537   
3  Sarojini Nagar              28.574157                77.19537   
4  Sarojini Nagar              28.574157                77.19537   

                        Venue  Venue Latitude  Venue Longitude  \
0              Domino's Pizza       28.576000        77.195000   
1       Sarojini Nagar Market       28.577802        77.196347   
2                  McDonald's       28.576515        77.196100   
3  Rdbd Rang de Basanti Dhaba       28.576552        77.195252   
4                  Haldiram's       28.576374        77.195266   

         Venue Category  
0           Pizza Place  
1                Market  
2  Fast Food Restaurant  
3     Indian Restaurant  
4     Indian Restaurant

### Transforming the data using one hot encoding

In [290]:
#one hot encoding
predict_onehot = pd.get_dummies(predict_venues[['Venue Category']], prefix="", prefix_sep="")
# add neighborhood column back to dataframe
predict_onehot['Neighborhood'] = predict_venues['Neighborhood'] 
# move neighborhood column to the first column
fixed_columns = [predict_onehot.columns[-1]] + list(predict_onehot.columns[:-1])
predict_onehot = predict_onehot[fixed_columns]
predict_onehot.head()

Neighborhood  Department Store  Dessert Shop  Fast Food Restaurant  \
0  Sarojini Nagar                 0             0                     0   
1  Sarojini Nagar                 0             0                     0   
2  Sarojini Nagar                 0             0                     1   
3  Sarojini Nagar                 0             0                     0   
4  Sarojini Nagar                 0             0                     0   

   Indian Restaurant  Market  Pizza Place  Shopping Mall  Women's Store  
0                  0       0            1              0              0  
1                  0       1            0              0              0  
2                  0       0            0              0              0  
3                  1       0            0              0              0  
4                  1       0            0              0              0

#### Ensuring labels are same for test & model data

In [291]:
columns=bangalore_onehot.columns
def add_missing_dummy_columns( d, columns ):
    missing_cols = set( columns ) - set( d.columns )
    for c in missing_cols:
        d[c] = 0
        
def fix_columns( d, columns ):  
    add_missing_dummy_columns( d, columns )
    # make sure we have all the columns we need
    assert( set( columns ) - set( d.columns ) == set())
    extra_cols = set( d.columns ) - set( columns )
    if extra_cols:
        print ("extra columns:", extra_cols)
    d = d[ columns ]
    return d

In [292]:

predict_onehot=fix_columns(predict_onehot,columns)
predict_onehot.head()


Neighborhood  ATM  Accessories Store  Afghan Restaurant  Airport Service  \
0  Sarojini Nagar    0                  0                  0                0   
1  Sarojini Nagar    0                  0                  0                0   
2  Sarojini Nagar    0                  0                  0                0   
3  Sarojini Nagar    0                  0                  0                0   
4  Sarojini Nagar    0                  0                  0                0   

   American Restaurant  Andhra Restaurant  Arcade  Art Gallery  \
0                    0                  0       0            0   
1                    0                  0       0            0   
2                    0                  0       0            0   
3                    0                  0       0            0   
4                    0                  0       0            0   

   Arts & Crafts Store  ...  Toy / Game Store  Track  Trail  Train Station  \
0                    0  ...                 0      0      0              0   
1                    0  ...                 0      0      0              0   
2                    0  ...                 0      0      0              0   
3                    0  ...                 0      0      0              0   
4                    0  ...                 0      0      0              0   

   Udupi Restaurant  Vegetarian / Vegan Restaurant  Vietnamese Restaurant  \
0                 0                              0                      0   
1                 0                              0                      0   
2                 0                              0                      0   
3                 0                              0                      0   
4                 0                              0                      0   

   Waterfront  Wine Shop  Women's Store  
0           0          0              0  
1           0          0              0  
2           0          0              0  
3           0          0              0  
4           0          0              0  

[5 rows x 166 columns]

### Predicting the cluster label using kmeans model

In [293]:
predict_grouped = predict_onehot.groupby('Neighborhood').mean().reset_index()
predict_grouped = predict_grouped.drop('Neighborhood', 1)
cluster_group=kmeans.predict(predict_grouped)
cluster_group
print(cluster_group[0])

2


### Plotting the clusters on map

In [294]:
destination_location=bangalore_merged.loc[bangalore_merged['Cluster Labels'] ==cluster_group[0]]
destination_location

Pincode                                       Neighborhood   Latitude  \
1    560002  Bangalore City S.O,Bangalore Corporation Build...  12.958625   
3    560004  Basavanagudi H.O,Mavalli S.O,Pampamahakavi Roa...  12.944829   
8    560009     Bangalore Dist Offices Bldg S.O,K. G. Road S.O  12.979649   
9    560010  Industrial Estate S.O (Bangalore),Rajajinagar ...  12.991038   
12   560013                                      Jalahalli H.O  13.052544   
16   560017                             NAL S.O,Vimanapura S.O  12.953824   
17   560018                         Chamrajpet S.O (Bangalore)  12.960327   
20   560021                  Gayathrinagar S.O,Srirampuram S.O  12.993840   
21   560022            Yeshwanthpur Bazar S.O,Yeswanthpura S.O  13.024783   
23   560024  Anandnagar S.O (Bangalore),H.A. Farm S.O,Hebba...  13.036578   
26   560027  Sampangiramnagar S.O,Shanthinagar S.O,Wilson G...  12.960414   
27   560029               Dharmaram College S.O,Taverekere B.O  12.926182   
30   560033            Malkand Lines S.O,Maruthi Sevanagar S.O  13.010777   
32   560035                                     Carmelaram S.O  12.953839   
35   560038                        Indiranagar S.O (Bangalore)  12.977318   
37   560040  Chandra Lay Out S.O,Vijayanagar East S.O,Vijay...  12.975579   
38   560041           Jayanagar H.O,Tilaknagar S.O (Bangalore)  12.925346   
39   560042                           Sivan Chetty Gardens S.O  12.980207   
40   560043  Banaswadi S.O,Horamavu B.O,Jalavayuvihar S.O,K...  13.028293   
41   560045  Arabic College S.O,Nagawara B.O,Venkateshapura...  13.020721   
47   560051                                    H.K.P. Road S.O  12.986216   
48   560053                                       Chickpet S.O  12.970289   
51   560056  Bnagalore Viswavidalaya S.O,Mallathahalli B.O,...  12.957204   
52   560057                             Peenya Dasarahalli S.O  13.042241   
56   560061              Chikkalasandra S.O,Subramanyapura S.O  12.909293   
61   560066  Immedihalli B.O,Ramagondana Halli B.O,Whitefie...  12.953561   
63   560068  Begur B.O,Bommanahalli S.O (Bangalore),Hongasa...  12.905870   
65   560071                                         Domlur S.O  12.957564   
68   560074  Kumbalagodu S.O,Kumbalgodu Gollahalli B.O,Ramo...  12.910079   
73   560079  Basaveshwaranagar S.O,K H B Colony S.O,Kamaksh...  12.985886   
74   560080                                 Sadashivanagar S.O  13.000095   
76   560083  Bannerghatta S.O,Gottigere S.O,Kallubalu B.O,R...  12.884159   
77   560084  Kacharakanahalli B.O,Lingarajapuram S.O,St. Th...  13.007577   
79   560086  Mahalakshmipuram Layout S.O,West of Chord Road...  12.999170   
83   560092  Amruthahalli B.O,Byatarayanapura B.O,Kodigehal...  13.060251   
84   560093                                C.V.Raman Nagar S.O  12.983166   
85   560094  ISRO Anthariksha Bhavan S.O,R.M.V. Extension I...  13.037151   
86   560095                              Koramangala VI Bk S.O  12.937529   
92   560102                                     HSR Layout S.O  13.008228   
94   560104                                     Hampinagar S.O  12.962577   

    Longitude  Cluster Labels          1st Most Common Venue  \
1   77.577567             2.0              Indian Restaurant   
3   77.567095             2.0              Indian Restaurant   
8   77.577441             2.0              Indian Restaurant   
9   77.549685             2.0              Indian Restaurant   
12  77.549532             2.0              Indian Restaurant   
16  77.656158             2.0              Indian Restaurant   
17  77.570430             2.0          General Entertainment   
20  77.555621             2.0              Indian Restaurant   
21  77.545966             2.0           Fast Food Restaurant   
23  77.596340             2.0              Indian Restaurant   
26  77.592187             2.0  Vegetarian / Vegan Restaurant   
27  77.597373             2.0              Indian Restaurant   
30  77.635051 

In [295]:
import matplotlib.cm as cm
import matplotlib.colors as colors
import folium 

# create map of Manhattan using latitude and longitude values
map_to = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(destination_location['Latitude'], destination_location['Longitude'], destination_location['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_to)  
    
map_to

According to our model the if the user stay at Sarojini Nagar,Delhi the similar neighborhood belongs to cluster 1.So the users can choose any of the above place.It is shown in map also. The pincodes where the person can stay is:

In [296]:
result=bangalore_merged.loc[bangalore_merged['Cluster Labels'] == cluster_group[0], bangalore_merged.columns[[1] + list(range(5, bangalore_merged.shape[1]))]]
result

Neighborhood  \
1   Bangalore City S.O,Bangalore Corporation Build...   
3   Basavanagudi H.O,Mavalli S.O,Pampamahakavi Roa...   
8      Bangalore Dist Offices Bldg S.O,K. G. Road S.O   
9   Industrial Estate S.O (Bangalore),Rajajinagar ...   
12                                      Jalahalli H.O   
16                             NAL S.O,Vimanapura S.O   
17                         Chamrajpet S.O (Bangalore)   
20                  Gayathrinagar S.O,Srirampuram S.O   
21            Yeshwanthpur Bazar S.O,Yeswanthpura S.O   
23  Anandnagar S.O (Bangalore),H.A. Farm S.O,Hebba...   
26  Sampangiramnagar S.O,Shanthinagar S.O,Wilson G...   
27               Dharmaram College S.O,Taverekere B.O   
30            Malkand Lines S.O,Maruthi Sevanagar S.O   
32                                     Carmelaram S.O   
35                        Indiranagar S.O (Bangalore)   
37  Chandra Lay Out S.O,Vijayanagar East S.O,Vijay...   
38           Jayanagar H.O,Tilaknagar S.O (Bangalore)   
39                           Sivan Chetty Gardens S.O   
40  Banaswadi S.O,Horamavu B.O,Jalavayuvihar S.O,K...   
41  Arabic College S.O,Nagawara B.O,Venkateshapura...   
47                                    H.K.P. Road S.O   
48                                       Chickpet S.O   
51  Bnagalore Viswavidalaya S.O,Mallathahalli B.O,...   
52                             Peenya Dasarahalli S.O   
56              Chikkalasandra S.O,Subramanyapura S.O   
61  Immedihalli B.O,Ramagondana Halli B.O,Whitefie...   
63  Begur B.O,Bommanahalli S.O (Bangalore),Hongasa...   
65                                         Domlur S.O   
68  Kumbalagodu S.O,Kumbalgodu Gollahalli B.O,Ramo...   
73  Basaveshwaranagar S.O,K H B Colony S.O,Kamaksh...   
74                                 Sadashivanagar S.O   
76  Bannerghatta S.O,Gottigere S.O,Kallubalu B.O,R...   
77  Kacharakanahalli B.O,Lingarajapuram S.O,St. Th...   
79  Mahalakshmipuram Layout S.O,West of Chord Road...   
83  Amruthahalli B.O,Byatarayanapura B.O,Kodigehal...   
84                                C.V.Raman Nagar S.O   
85  ISRO Anthariksha Bhavan S.O,R.M.V. Extension I...   
86                              Koramangala VI Bk S.O   
92                                     HSR Layout S.O   
94                                     Hampinagar S.O   

            1st Most Common Venue          2nd Most Common Venue  \
1               Indian Restaurant                  Historic Site   
3               Indian Restaurant           Fast Food Restaurant   
8               Indian Restaurant                          Hotel   
9               Indian Restaurant                         Bakery   
12              Indian Restaurant           Fast Food Restaurant   
16              Indian Restaurant                     Restaurant   
17          General Entertainment              Indian Restaurant   
20              Indian Restaurant                         Bakery   
21           Fast Food Restaurant             Miscellaneous Shop   
23              Indian Restaurant               Department Store   
26  Vegetarian / Vegan Restaurant              Indian Restaurant   
27              Indian Restaurant               Asian Restaurant   
30              Indian Restaurant              Andhra Restaurant   
32              Indian Restaurant                  Women's Store   
35              Indian Restaurant                           Café   
37              Indian Restaurant                    Bus Station   
38              Indian Restaurant               Department Store   
39              Indian Restaurant                           Café   
40              Indian Restaurant      Indian Chinese Restaurant   
41              Indian Restaurant                 Clothing Store   
47              Indian Restaurant                           Café   
48              Indian Restaurant             Miscellaneous Shop   
51                            ATM              Indian Restaurant   
52              Indian Restaurant                    Bus Station   
56 

Just for comparison the top 10 venue nearby user location is :

In [297]:
predict_venues

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0   Sarojini Nagar              28.574157                77.19537   
1   Sarojini Nagar              28.574157                77.19537   
2   Sarojini Nagar              28.574157                77.19537   
3   Sarojini Nagar              28.574157                77.19537   
4   Sarojini Nagar              28.574157                77.19537   
5   Sarojini Nagar              28.574157                77.19537   
6   Sarojini Nagar              28.574157                77.19537   
7   Sarojini Nagar              28.574157                77.19537   
8   Sarojini Nagar              28.574157                77.19537   
9   Sarojini Nagar              28.574157                77.19537   
10  Sarojini Nagar              28.574157                77.19537   
11  Sarojini Nagar              28.574157                77.19537   

                              Venue  Venue Latitude  Venue Longitude  \
0                    Domino's Pizza       28.576000        77.195000   
1             Sarojini Nagar Market       28.577802        77.196347   
2                        McDonald's       28.576515        77.196100   
3        Rdbd Rang de Basanti Dhaba       28.576552        77.195252   
4                        Haldiram's       28.576374        77.195266   
5                        McDonald's       28.575960        77.195201   
6                 South Square Mall       28.576386        77.195325   
7                       Babu Market       28.577409        77.195647   
8   South Square Multilevel Parking       28.577434        77.195623   
9                     Khushi Sarees       28.577427        77.195627   
10                Harajuku Delights       28.577602        77.194353   
11                      Kulcha King       28.569892        77.194325   

          Venue Category  
0            Pizza Place  
1                 Market  
2   Fast Food Restaurant  
3      Indian Restaurant  
4      Indian Restaurant  
5   Fast Food Restaurant  
6          Shopping Mall  
7          Women's Store  
8       Department Store  
9          Women's Store  
10          Dessert Shop  
11  Fast Food Restaurant

In [298]:
venue_category=predict_venues['Venue Category'].tolist()
venue_count=list()
result.shape


for i in range(len(result)):
    score=0
    for j in range(11):
        if result.iloc[i,j] in venue_category : 
            score=score+1
        else:
            continue
    score=(score/10)
    venue_count.append(score)
   
    

In [299]:
venue_category=predict_venues['Venue Category'].tolist()
venue_count=list()
size = len(venue_category)

for i in range(len(result)):
    score=0
    for venue in venue_category:
        for j in range(result.shape[1]):
            if result.iloc[i,j] == venue :
                score=score+1
                break;
    score=(score/size)
    venue_count.append(score)

In [300]:
print(venue_category)

['Pizza Place', 'Market', 'Fast Food Restaurant', 'Indian Restaurant', 'Indian Restaurant', 'Fast Food Restaurant', 'Shopping Mall', "Women's Store", 'Department Store', "Women's Store", 'Dessert Shop', 'Fast Food Restaurant']


In [301]:
result['accuracy']=venue_count
result

Neighborhood  \
1   Bangalore City S.O,Bangalore Corporation Build...   
3   Basavanagudi H.O,Mavalli S.O,Pampamahakavi Roa...   
8      Bangalore Dist Offices Bldg S.O,K. G. Road S.O   
9   Industrial Estate S.O (Bangalore),Rajajinagar ...   
12                                      Jalahalli H.O   
16                             NAL S.O,Vimanapura S.O   
17                         Chamrajpet S.O (Bangalore)   
20                  Gayathrinagar S.O,Srirampuram S.O   
21            Yeshwanthpur Bazar S.O,Yeswanthpura S.O   
23  Anandnagar S.O (Bangalore),H.A. Farm S.O,Hebba...   
26  Sampangiramnagar S.O,Shanthinagar S.O,Wilson G...   
27               Dharmaram College S.O,Taverekere B.O   
30            Malkand Lines S.O,Maruthi Sevanagar S.O   
32                                     Carmelaram S.O   
35                        Indiranagar S.O (Bangalore)   
37  Chandra Lay Out S.O,Vijayanagar East S.O,Vijay...   
38           Jayanagar H.O,Tilaknagar S.O (Bangalore)   
39                           Sivan Chetty Gardens S.O   
40  Banaswadi S.O,Horamavu B.O,Jalavayuvihar S.O,K...   
41  Arabic College S.O,Nagawara B.O,Venkateshapura...   
47                                    H.K.P. Road S.O   
48                                       Chickpet S.O   
51  Bnagalore Viswavidalaya S.O,Mallathahalli B.O,...   
52                             Peenya Dasarahalli S.O   
56              Chikkalasandra S.O,Subramanyapura S.O   
61  Immedihalli B.O,Ramagondana Halli B.O,Whitefie...   
63  Begur B.O,Bommanahalli S.O (Bangalore),Hongasa...   
65                                         Domlur S.O   
68  Kumbalagodu S.O,Kumbalgodu Gollahalli B.O,Ramo...   
73  Basaveshwaranagar S.O,K H B Colony S.O,Kamaksh...   
74                                 Sadashivanagar S.O   
76  Bannerghatta S.O,Gottigere S.O,Kallubalu B.O,R...   
77  Kacharakanahalli B.O,Lingarajapuram S.O,St. Th...   
79  Mahalakshmipuram Layout S.O,West of Chord Road...   
83  Amruthahalli B.O,Byatarayanapura B.O,Kodigehal...   
84                                C.V.Raman Nagar S.O   
85  ISRO Anthariksha Bhavan S.O,R.M.V. Extension I...   
86                              Koramangala VI Bk S.O   
92                                     HSR Layout S.O   
94                                     Hampinagar S.O   

            1st Most Common Venue          2nd Most Common Venue  \
1               Indian Restaurant                  Historic Site   
3               Indian Restaurant           Fast Food Restaurant   
8               Indian Restaurant                          Hotel   
9               Indian Restaurant                         Bakery   
12              Indian Restaurant           Fast Food Restaurant   
16              Indian Restaurant                     Restaurant   
17          General Entertainment              Indian Restaurant   
20              Indian Restaurant                         Bakery   
21           Fast Food Restaurant             Miscellaneous Shop   
23              Indian Restaurant               Department Store   
26  Vegetarian / Vegan Restaurant              Indian Restaurant   
27              Indian Restaurant               Asian Restaurant   
30              Indian Restaurant              Andhra Restaurant   
32              Indian Restaurant                  Women's Store   
35              Indian Restaurant                           Café   
37              Indian Restaurant                    Bus Station   
38              Indian Restaurant               Department Store   
39              Indian Restaurant                           Café   
40              Indian Restaurant      Indian Chinese Restaurant   
41              Indian Restaurant                 Clothing Store   
47              Indian Restaurant                           Café   
48              Indian Restaurant             Miscellaneous Shop   
51                            ATM              Indian Restaurant   
52              Indian Restaurant                    Bus Station   
56 

In [305]:
result.shape

(40, 52)

In [302]:
acc=result['accuracy'].max()
print(acc)
highlight=result[result['accuracy']>=acc]
highlight.reset_index(inplace=True) # Resets the index, makes factor a column
highlight.iloc[0,1]
ht=fin_bangalore_data[fin_bangalore_data['PostOffice']==highlight.iloc[0,1]]
print(ht)
lt=ht.iloc[0,2]
ln=ht.iloc[0,3]

0.9166666666666666
    Pincode                                         PostOffice   Latitude  \
23   560024  Anandnagar S.O (Bangalore),H.A. Farm S.O,Hebba...  13.036578   

    Longitude  
23   77.59634  


In [303]:
print(lt)
print(ln)

13.0365775420601
77.59633976975934


In [304]:
import matplotlib.cm as cm
import matplotlib.colors as colors
import folium 

# create map of Manhattan using latitude and longitude values
map_to = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(destination_location['Latitude'], destination_location['Longitude'], destination_location['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_to)  
    folium.Marker( [lt,ln] ).add_to( map_to )
   

map_to

Great! so  we have found most simliar neighborhood Anandnagar S.O (Bangalore),H.A. Farm S.O,Hebba with Similarity score 91%.

## Results and Discussion <a name="results"></a>

In our analysis  we have taken a demo input let's say the person living nearby Sarojini Nagar ,Delhi(PinCode-110023) and willing to relocate to bangalore we got all the nearby venue of Sarojini Nagar in 500m radius using FoursquareAPI. It is going to work with any user provided input.

We have taken all the pincodes of bangalore  as data point and found the nearby 50 venues withinh 500m of those pincodes.
Those location candidates were then clustered in 5 groups to create zones of interest which contains similar kind of neighborhood. All the pincodes are labelled on the basis of their cluster group. Considering our demo case, we found the zone areas(cluster group 1) which is having neighborhood similiar to Sarojini Nagar,Delhi.After identifying the clusters we tried to find the similarity score(accuracy)for each area(pincode) in the given cluster on the basis of percent of venues matched in the user location and predicted areas. 
On the basis of Similarity scores we have highlighted the area having maximum similarity.

Result of all this is 40 areas similar to the user location has been predicted alongwith their similarity score.We have found Anandnagar S.O (Bangalore),H.A. Farm S.O,Hebbal(560024) is very similar to Sarojini Nagar alongwith similarity of 91%.  This, of course, does not imply that those areas are actually optimal locations for a relocation! Purpose of this analysis was to only provide info on bangalore areas similar to the neighborhood of user area, but other factors can also be included like commuting facilities, distance from office,levels of noise etc.

## Conclusion <a name="conclusion"></a>

Purpose of this project was to identify the bangalore city areas similar to the neighborhood area of the user in another city,the information can be utilised for relocation purpose. By using the FoursquareAPI we clustered the bangalore areas on the basis of nearby venue(similar neighborhood) and then recommended the  simliar areas to the user alongwith similarity score.

Final decision on relocation will be made by users based on specific characteristics of neighborhoods and locations in every recommended zone, taking into consideration additional factors like attractiveness of each location (proximity to park or water), levels of noise / proximity to major roads, commuting facilities etc.